internal/zero/bootstrap/writers/k8s/secret.go

```diff
@@ -0,0 +1,176 @@
+package k8s
+
+import (
+	"context"
+	"crypto/cipher"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/pomerium/pomerium/internal/log"
+	"github.com/pomerium/pomerium/internal/zero/bootstrap/writers"
+	"github.com/pomerium/pomerium/pkg/cryptutil"
+	cluster_api "github.com/pomerium/pomerium/pkg/zero/cluster"
+)
+
+func init() {
+	writers.RegisterBuilder("secret", func(uri *url.URL) (writers.ConfigWriter, error) {
+		return newSecretWriter(uri)
+	})
+}
+
+type secretWriter struct {
+	client       *http.Client
+	apiserverURL *url.URL
+	namespace    string
+	name         string
+	key          string
+}
+
+func newSecretWriter(uri *url.URL) (*secretWriter, error) {
+	client, apiserverURL, err := inClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	parts := strings.SplitN(path.Join(uri.Host, uri.Path), "/", 3)
+	if len(parts) != 3 || parts[1] == "" || parts[2] == "" {
+		return nil, fmt.Errorf("invalid secret uri %q, expecting format \"secret://namespace/name/key\"", uri.String())
+	} else if parts[0] == "" {
+		return nil, fmt.Errorf(`invalid secret uri %q (did you mean "secret:/%s"?)`, uri.String(), uri.Path)
+	}
+	return &secretWriter{
+		client:       client,
+		apiserverURL: apiserverURL,
+		namespace:    parts[0],
+		name:         parts[1],
+		key:          parts[2],
+	}, nil
+}
+
+// WriteConfig implements ConfigWriter.
+func (w *secretWriter) WriteConfig(ctx context.Context, src *cluster_api.BootstrapConfig, cipher cipher.AEAD) error {
+	u := w.apiserverURL.ResolveReference(&url.URL{
+		Path: path.Join("/api/v1/namespaces", w.namespace, "secrets", w.name),
+		RawQuery: url.Values{
+			"fieldManager": {"pomerium-zero"},
+			"force":        {"true"},
+		}.Encode(),
+	})
+	plaintext, err := json.Marshal(src)
+	if err != nil {
+		return err
+	}
+	ciphertext := cryptutil.Encrypt(cipher, plaintext, nil)
```

Comment:
Could go either way I think. It may be easier to encrypt it so that it works the same regardless of writer backend. Depends if we want this to be user-serviceable or not. If not, having the data be raw bytes without a visible encoding might help signal that this is not meant to be hand-edited.